lightningd/jsonrpc.c

```diff
@@ -633,22 +700,27 @@ parse_request(struct json_connection *jcon, const jsmntok_t tok[])
 				    "Expected string for method");
 	}
 
-	c->json_cmd = find_cmd(jcon->ld->jsonrpc, jcon->buffer, method);
+	c->json_cmd = find_cmd(ld->jsonrpc, buffer, method);
 	if (!c->json_cmd) {
 		return command_fail(
 		    c, JSONRPC2_METHOD_NOT_FOUND, "Unknown command '%.*s'",
-		    method->end - method->start, jcon->buffer + method->start);
+		    method->end - method->start, buffer + method->start);
 	}
 	if (c->json_cmd->deprecated && !deprecated_apis) {
 		return command_fail(c, JSONRPC2_METHOD_NOT_FOUND,
 				    "Command '%.*s' is deprecated",
 				    method->end - method->start,
-				    jcon->buffer + method->start);
+				    buffer + method->start);
 	}
 
-	db_begin_transaction(jcon->ld->wallet->db);
-	res = c->json_cmd->dispatch(c, jcon->buffer, tok, params);
-	db_commit_transaction(jcon->ld->wallet->db);
+	/* Especially for internal call, if we set a timer, and the timer loop
+	 * of `lightningd` will make the timer callback in db transaction. */
+	if (!db_already_in_transaction(ld->wallet->db)) {
+		db_begin_transaction(ld->wallet->db);
+		res = c->json_cmd->dispatch(c, buffer, tok, params);
+		db_commit_transaction(ld->wallet->db);
+	} else
+		res = c->json_cmd->dispatch(c, buffer, tok, params);
```

Comment:
I don't think this can work, at least in this form.

Nesting transactions is always a bad idea.  Various callbacks we have assume that the transaction will be successfully completed before anything else happens: e.g. we'll commit to disk, then send something to a subdaemon.  This implicit handling keeps our data consistency model simple (and, frankly, managable!).

I'm still reviewing the patches, and am having trouble seeing the forest through the trees so far...

Reply:
`getclientversion` will be called before rpc initials and I want to use timer to retry `getclientversion`, but `io_loop_with_timers` always use the transaction which will cause the conflict with here.
I think the alternative way is to implement the waiter like `struct sync_waiter`. What about implementing the waiter to pending all request?